### PR TITLE
WIP: Introduce CoinGecko API Endpoints

### PR DESCRIPTION
**Pull request by dpatsora**

### CODE REVIEW

File: app/api/v2/coin_gecko/entities/pair.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      module Entities
+        class Pair < API::V2::Entities::Base
+          expose(
+              :ticker_id,
+              documentation: {
+                  type: String,
+                  desc: 'A market ID with delimiter.'
+              }
+          ) do |market|
+            "#{market[:base_unit].upcase}_#{market[:quote_unit].upcase}"
```
Comment (mnaichuk) [inline app/api/v2/coin_gecko/entities/pair.rb]: can be the helper in the market model

File: app/api/v2/coin_gecko/entities/pair.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+          end
+
+          expose(
+              :base,
+              documentation: {
+                  type: String,
+                  desc: 'A currency code of the base asset.'
+              }
+          ) do |market|
+            market[:base_unit].upcase
```
Comment (mnaichuk) [inline app/api/v2/coin_gecko/entities/pair.rb:25]: can be the helper in the market model

File: app/api/v2/coin_gecko/entities/pair.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+          end
+
+          expose(
+              :target,
+              documentation: {
+                  type: String,
+                  desc: 'A currency code of the quote asset.'
+              }
+          ) do |market|
+            market[:quote_unit].upcase
```
Comment (mnaichuk) [inline app/api/v2/coin_gecko/entities/pair.rb:35]: can be the helper in the market model

File: app/api/v2/coin_gecko/helpers.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+module API
+  module V2
+    module CoinGecko
+      module Helpers
+        def format_ticker(ticker, market)
+          # FIXME: Need to add bid and ask
+          {
+              ticker_id: "#{market[:base_unit].upcase}_#{market[:quote_unit].upcase}",
```
Comment (mnaichuk) [inline app/api/v2/coin_gecko/helpers.rb]: can be the helpers in the market model

File: app/api/v2/coin_gecko/pairs.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
+# encoding: UTF-8
+# frozen_string_literal: true
+
+module API
+  module V2
+    module CoinGecko
+      class Pairs < Grape::API
+        desc 'Get list of all available trading pairs'
+        get "/pairs" do
+          enabled_markets = ::Market.enabled
```
Comment (mnaichuk) [inline app/api/v2/coin_gecko/pairs.rb]: we can use redis here

File: app/api/v2/coin_gecko/tickers.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+module API
+  module V2
+    module CoinGecko
+      class Tickers < Grape::API
+        desc 'Get list of all available trading pairs'
+        get "/tickers" do
+          enabled_markets = ::Market.enabled.ordered
+
+          tickers = enabled_markets.map do |market|
+            format_ticker(TickersService[market].ticker, market)
```
Comment (mnaichuk) [inline app/api/v2/coin_gecko/tickers.rb]: we can use redis here

**Review by mnaichuk — CHANGES_REQUESTED** [review]

File: app/api/v2/coin_gecko/historical_trades.rb [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+          if params[:type].present?
+            trades_query += " AND taker_type=%{type}"
+          end
+
+          if params[:start_time].present?
+            trades_query += " AND created_at>=%{start_time}"
+          end
+
+          if params[:end_time].present?
+            trades_query += " AND created_at>=%{end_time}"
```
Comment (mnaichuk) [inline app/api/v2/coin_gecko/historical_trades.rb]: fix this please

File: app/api/v2/coin_gecko/historical_trades.rb [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+          end
+
+          trades_query += " ORDER BY desc"
+
+          unless params[:limit].to_d.zero?
```
Comment (mnaichuk) [inline app/api/v2/coin_gecko/historical_trades.rb]: I'm 100% sure that we should have some limit or we should have a limitation between start_at and end_at

File: app/api/v2/coin_gecko/historical_trades.rb [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+            trades_query += " LIMIT %{limit}"
+          end
+
+          Trade.public_from_influx_with_filters(market.id, trades_query, filters).each do |trade|
```
Comment (mnaichuk) [inline app/api/v2/coin_gecko/historical_trades.rb]: each_with_object({'buy' => [], 'sell' => []})

**Review by calj — APPROVED** [review]

**Review by mnaichuk — APPROVED** [review]

**Pull request merged**